File: src/exoticatechnologies/modifications/exotics/ETExotics.java
package exoticatechnologies.modifications.exotics;

import lombok.extern.log4j.Log4j;

import java.util.*;

@Log4j
public class ETExotics {
    private List<String> exotics = null;
    private final Map<String, ExoticData> exoticData = new HashMap<>();

    public ETExotics() {
        fixExoticsList();
    }

    public ETExotics(List<String> exotics) {
        for(String key : exotics) {
            this.putExotic(key);
        }
    }

    public Collection<String> getList() {
        fixExoticsList();
        return new ArrayList<>(exoticData.keySet());
    }

    public ExoticData getData(Exotic exotic) {
        return exoticData.get(exotic.getKey());
    }

    public boolean hasExotic(Exotic exotic) {
        return this.hasExotic(exotic.getKey());
    }

    public boolean hasExotic(String key) {
        fixExoticsList();
        return this.exoticData.containsKey(key);
    }

    public void putExotic(Exotic exotic) {
        this.putExotic(exotic.getKey());
    }

    public void putExotic(String key) {
        this.exoticData.put(key, new ExoticData(key));
    }

    public void removeExotic(String key) {
        this.exoticData.remove(key);
    }

    public void removeExotic(Exotic exotic) {
        this.exoticData.remove(exotic.getKey());
    }

    public boolean hasAnyExotic() {
        return !this.exoticData.isEmpty();
    }

    public List<String> getTags() {
        Set<String> tagSet = new HashSet<>();
        for (String key : exoticData.keySet()) {
            Exotic exotic = ExoticsHandler.EXOTICS.get(key);
            if (exotic.getTag() != null) {
                tagSet.add(exotic.getTag());
            }
        }

        List<String> tags = new ArrayList<>(tagSet);
        return tags;
    }

    public List<Exotic> getConflicts(String tag) {
        List<Exotic> exotics = new ArrayList<>();
        for (String key : this.exoticData.keySet()) {
            Exotic exotic = ExoticsHandler.EXOTICS.get(key);
            if (exotic.getTag() != null && exotic.getTag().equals(tag)) {
                exotics.add(exotic);
            }
        }
        return exotics;
    }

    public void fixExoticsList() {
        if (this.exotics != null) {
            for (String exoticKey : this.exotics) {
                this.exoticData.put(exoticKey, new ExoticData(exoticKey));
            }
            this.exotics = null;
        }
    }

    @Override
    public String toString() {
        return "ETExotics{" +
                "exotics=" + exotics +
                '}';
    }
}

File: src/exoticatechnologies/modifications/exotics/ETExotics.kt
package exoticatechnologies.modifications.exotics

import com.fs.starfarer.api.fleet.FleetMemberAPI
import lombok.extern.log4j.Log4j
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

@Log4j
class ETExotics {
    var exotics: List<String>? = null
    var exoticData: MutableMap<String, ExoticData> = HashMap()

    constructor() {
        fixExoticsList()
    }

    constructor(exotics: List<String>) {
        for (key in exotics) {
            this.putExotic(key)
        }
    }

    val list: Collection<String>
        get() {
            fixExoticsList()
            return ArrayList(exoticData.keys)
        }

    fun getData(exotic: Exotic): ExoticData? {
        return exoticData[exotic.key]
    }

    fun getCount(member: FleetMemberAPI): Int {
        return exoticData
            .filterValues { it.exotic.countsTowardsExoticLimit(member) }
            .count()
    }

    fun hasExotic(exotic: Exotic): Boolean {
        return this.hasExotic(exotic.key)
    }

    fun hasExotic(key: String): Boolean {
        fixExoticsList()
        return exoticData.containsKey(key)
    }

    fun putExotic(exoticData: ExoticData) {
        this.exoticData[exoticData.key] = exoticData
    }

    fun putExotic(key: String?) {
        putExotic(ExoticData(key!!))
    }

    fun removeExotic(key: String) {
        exoticData.remove(key)
    }

    fun removeExotic(exotic: Exotic) {
        exoticData.remove(exotic.key)
    }

    fun hasAnyExotic(): Boolean {
        fixExoticsList()
        return exoticData.isNotEmpty()
    }

    val tags: List<String>
        get() {
            val tagSet: MutableSet<String> = HashSet()
            for (data in exoticData.values) {
                val exotic = data.exotic
                tagSet.addAll(exotic.tags)
            }
            return ArrayList(tagSet)
        }

    fun getConflicts(tags: List<String>): List<Exotic> {
        val exotics: MutableList<Exotic> = ArrayList()
        for (data in exoticData.values) {
            val exotic = data.exotic
            exotic.tags
                .firstOrNull { tags.contains(it) }
                ?.let {
                    exotics.add(exotic)
                }
        }
        return exotics
    }

    fun fixExoticsList() {
        if (exotics != null) {
            for (exoticKey in exotics!!) {
                exoticData.getOrPut(exoticKey){ ExoticData(exoticKey) }
            }
            exotics = null
        }
    }

    override fun toString(): String {
        return "ETExotics{" +
                "exoticData=" + exoticData +
                '}'
    }

    fun toJson(): JSONObject? {
        val obj = JSONObject()
        exoticData.values.forEach {
            obj.put(it.key, it.toJson())
        }
        return obj
    }

    @Throws(JSONException::class)
    fun parseJson(obj: Any) {
        if (obj is JSONArray) {
            parseJsonArray(obj)
        } else if (obj is JSONObject) {
            parseJsonObj(obj)
        }
    }

    private fun parseJsonObj(obj: JSONObject) {
        for (key in obj.keys()) {
            val exoObj = obj.getJSONObject(key.toString())
            putExotic(ExoticData(exoObj))
        }
    }

    private fun parseJsonArray(obj: JSONArray) {
        for (i in 0 until obj.length()) {
            putExotic(obj.getString(i))
        }
    }
}